smartsim/entity/_new_ensemble.py

```diff
@@ -0,0 +1,93 @@
+# BSD 2-Clause License
+#
+# Copyright (c) 2021-2024, Hewlett Packard Enterprise
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import copy
+import itertools
+import os
+import typing as t
+
+# TODO: We are using the mocks defined `mock` module. We should use the correct
+#       module when these types are actually defined when PRs #587 and #603 are
+#       merged.
+# https://github.com/CrayLabs/SmartSim/pull/587
+# https://github.com/CrayLabs/SmartSim/pull/603
+from smartsim.entity import _mock, entity, strategies
+from smartsim.entity.files import EntityFiles
+from smartsim.entity.model import Application
+
+
+# TODO: If we like this design, we need to:
+#         1) Move this to the `smartsim._core.entity.ensemble` module
+#         2) Decide what to do witht the original `Ensemble` impl
+class Ensemble(entity.CompoundEntity):
```

Comment:
Is this class replacing the original Ensemble class?

Reply:
That's the plan! I just didn't want to remove the original yet in case anyone was using it for reference. I'd be more than willing to move this implementation to `smartsim.entity.ensemble`, delete the `# TODO` comment and remove the original if that's okay with you!